crates/ra_proc_macro/src/msg.rs

```diff
@@ -0,0 +1,218 @@
+//! A simplified version of lsp base protocol for rpc
+
+use std::{
+    fmt,
+    io::{self, BufRead, Write},
+};
+
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(untagged)]
+pub enum Message {
+    Request(Request),
+    Response(Response),
+}
+
+impl From<Request> for Message {
+    fn from(request: Request) -> Message {
+        Message::Request(request)
+    }
+}
+
+impl From<Response> for Message {
+    fn from(response: Response) -> Message {
+        Message::Response(response)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[serde(transparent)]
+pub struct RequestId(IdRepr);
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[serde(untagged)]
+enum IdRepr {
+    U64(u64),
+    String(String),
+}
+
+impl From<u64> for RequestId {
+    fn from(id: u64) -> RequestId {
+        RequestId(IdRepr::U64(id))
+    }
+}
+
+impl From<String> for RequestId {
+    fn from(id: String) -> RequestId {
+        RequestId(IdRepr::String(id))
+    }
+}
+
+impl fmt::Display for RequestId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            IdRepr::U64(it) => fmt::Display::fmt(it, f),
+            IdRepr::String(it) => fmt::Display::fmt(it, f),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Request {
+    pub id: RequestId,
+    pub method: String,
+    pub params: serde_json::Value,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Response {
+    // JSON RPC allows this to be null if it was impossible
+    // to decode the request's id. Ignore this special case
+    // and just die horribly.
+    pub id: RequestId,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub result: Option<serde_json::Value>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub error: Option<ResponseError>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ResponseError {
+    pub code: i32,
+    pub message: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data: Option<serde_json::Value>,
+}
+
+#[derive(Clone, Copy, Debug)]
+#[allow(unused)]
```

Comment:
Should we use `#[repr(u32)]` to be able to reliably cast it to `u32`?

Reply:
We don't cast it to `u32` anywhere. IIUC, these numbers are defined in jsonrpc spec such that make them following the spec. And all code in `msg.rs` I just copy and paste from https://github.com/rust-analyzer/lsp-server/blob/master/src/msg.rs

Reply:
As a curiosity, the request id is not number:

> An identifier established by the Client that MUST contain a String, Number, or NULL value if included. If it is not included it is assumed to be a notification. The value SHOULD normally not be Null [1] and Numbers SHOULD NOT contain fractional parts [2]

Reply:
We do it in `client_loop()` [here](https://github.com/rust-analyzer/rust-analyzer/pull/3738/files/8f42688ba101f08c8ca6404fbdf17e52bc55a8b8#diff-961389dd0d52a597fbd08bc9c18c1805R165) in this PR (I misslled, it is `i32` there instead).
Anyway, I wish we put a comment somewhere that this code is copy-pasted and the motivation for that (e.g. motivation against using the library it was copied from).